evm/src/main/java/org/hyperledger/besu/evm/code/CodeV1.java

```diff
@@ -38,6 +573,216 @@ public class CodeV1 implements Code {
+      return "Relative jump destinations targets invalid immediate data";
+    }
+    return null;
+  }
+
+  static String validateStack(final EOFLayout eofLayout) {
+    return IntStream.range(0, eofLayout.getCodeSections().length)
+        .mapToObj(i -> validateStack(i, eofLayout.getCodeSections()))
+        .filter(Objects::nonNull)
+        .findAny()
+        .orElse(null);
+  }
+
+  /**
+   * Validates the stack heights per <a href="https://eips.ethereum.org/EIPS/eip-5450">EIP-5450</a>.
+   *
+   * <p>This presumes code validation has already been performed, so there are no RJUMPS into
+   * immediates as well as no immediates falling off of the end of code sections.
+   *
+   * @param codeSectionToValidate The index of code to validate in the code sections
+   * @param codeSections The code sections to use for CALLF and JUMPF reference validation.
+   * @return null if valid, otherwise an error string providing the validation error.
+   */
+  public static String validateStack(
+      final int codeSectionToValidate, final CodeSection[] codeSections) {
+    byte[] code = codeSections[codeSectionToValidate].code.toArrayUnsafe();
+    int codeLength = code.length;
+    int[] stackHeights = new int[codeLength];
+    Arrays.fill(stackHeights, -1);
+
+    int thisWork = 0;
+    int maxWork = 1;
+    int[][] workList = new int[codeLength][2];
+
+    int initialStackHeight = codeSections[codeSectionToValidate].getInputs();
+    int maxStackHeight = initialStackHeight;
+    stackHeights[0] = initialStackHeight;
+    workList[0][1] = initialStackHeight;
+
+    while (thisWork < maxWork) {
+      int currentPC = workList[thisWork][0];
+      int currentStackHeight = workList[thisWork][1];
+      if (thisWork > 0 && stackHeights[currentPC] >= 0) {
+        // we've been here, validate the jump is what is expected
+        if (stackHeights[currentPC] != currentStackHeight) {
+          return String.format(
+              "Jump into code stack height (%d) does not match previous value (%d)",
+              stackHeights[currentPC], currentStackHeight);
+        } else {
+          thisWork++;
+          continue;
+        }
+      }
+
+      while (currentPC < codeLength) {
+        int thisOp = code[currentPC] & 0xff;
+        int recordedStack = stackHeights[currentPC];
+        if (recordedStack >= 0 && currentStackHeight != recordedStack) {
+          return String.format(
+              "Stack height mismatch %d/%d at %d", recordedStack, currentStackHeight, currentPC);
+        }
+
+        byte[] stackInfo = opcodeStackValidation[thisOp];
+        int stackInputs;
+        int stackOutputs;
+        int pcAdvance = stackInfo[2];
+        if (pcAdvance == 0) {
+          return String.format("Invalid opcode 0x%02X", thisOp);
+        }
+        if (thisOp == CallFOperation.OPCODE || thisOp == JumpFOperation.OPCODE) {
+          int section = ((code[currentPC + 1] & 0xff) << 8) | (code[currentPC + 2] & 0xff);
```

Comment:
Can use the shiny new Words.readBigEndianU16() here.

Reply:
Yea, this was written before the callf features merged.  will migrate.